### PR TITLE
fix(mitm-addon): bound auth base forwarded responses

### LINKED ISSUE #15065: security(mitm-addon): unbounded response body read in auth.base URL forwarder
## Summary

`auth_base_forwarder.py:186` reads the entire upstream response body into memory via `resp.read()` with no byte limit. This creates a memory-exhaustion denial-of-service vector for the runner-level mitmproxy process.

## Details

- **File:** `crates/runner/mitm-addon/src/auth_base_forwarder.py`
- **Line:** 186 - `resp_body = resp.read()`
- **Function:** `_forward_request_sync` (the sync helper used by `forward_request` for `auth.base` URL rewriting)

The connection timeout is 30 seconds, but that only bounds wall-clock time. It does not bound memory. A controlled upstream can still return a very large body within that window and force the addon to buffer it all before returning a response to the sandbox.

The proxy is runner-level: one mitmproxy process handles traffic for multiple sandbox VMs, with the proxy registry mapping source IPs to runs. A crash therefore affects more than the single request that triggered it.

## Failure scenario

1. A run uses a firewall entry with `auth.base` URL rewriting.
2. The resolved upstream returns a very large response body.
3. `_forward_request_sync` buffers the full body via `resp.read()`.
4. The runner-level mitmproxy process can run out of memory and crash.
5. Other sandbox traffic on the same runner can lose proxy service until mitmproxy restarts.
6. Any in-memory pending or buffered usage work in that process may be lost.

## Security impact

- **Runner-level DoS:** A user-controlled or compromised upstream can crash the shared runner mitmproxy process.
- **Noisy-neighbor risk:** Because mitmproxy is shared at the runner level, one run's oversized `auth.base` response can disrupt co-located sandbox traffic on that runner.
- **Usage-integrity risk:** A crash can lose process-local usage state that has not yet flushed. This is not currently a direct connector-specific billing bypass in the `auth.base` path: the current billable connector list is `x`, and `x` does not use `auth.base`. The risk is process-level usage loss on crash.

## Context

Most response consumers in the addon already have explicit bounds:

- `STREAM_BUFFER_LIMIT` = 64 KB for streaming response buffers.
- `LARGE_RESPONSE_DECOMPRESS_LIMIT` = 5 MB for selected decompression fallback paths.
- JSON selective extraction uses per-field `max_bytes` caps.

This low-level `auth.base` forwarder is the remaining unbounded full-response consumer.

Current generated firewall entries that use `auth.base` are small webhook-style rewrites plus DefiLlama-style API-key-in-path rewriting. DefiLlama is a real data API, so an overly small passthrough cap can break legitimate responses.

`forward_request()` offloads sync I/O with `asyncio.to_thread`, so limiting only one response body is not a complete process-level bound: several concurrent `auth.base` forwards can each allocate up to the per-response cap.

## Recommended fix

Use bounded full-response forwarding and cap concurrent `auth.base` forwarding work.

Implementation shape:

1. Add a dedicated `auth.base` forwarded response body limit, for example `MAX_AUTH_BASE_RESPONSE_BODY_BYTES = 32 * 1024 * 1024`. Do not reuse the 5 MB decompression limit unless we intentionally want that tighter product behavior for all `auth.base` APIs.
2. Replace the unbounded `resp.read()` with a single bounded read of `limit + 1` bytes.
3. If the result exceeds the limit, raise a specific forwarding exception such as `ForwardedResponseTooLargeError`.
4. Add a small `auth.base` forwarding concurrency cap, for example `MAX_CONCURRENT_AUTH_BASE_FORWARDS = 4`, around the `asyncio.to_thread(_forward_request_sync, ...)` call.
5. Let the existing `_apply_url_rewrite` exception path return the current 502 `url_rewrite_forward_failed` response.

This gives a predictable upper bound for response-body buffering from this path of roughly `MAX_AUTH_BASE_RESPONSE_BODY_BYTES * MAX_CONCURRENT_AUTH_BASE_FORWARDS`, while preserving the existing failure semantics.

Do not add a `Content-Length` precheck as part of the first fix. It is not a sufficient safety boundary because upstreams can omit or lie about it, and it adds HTTP semantic edge cases such as `HEAD`, `1xx`, `204`, and `304` responses. The bounded read is the required memory guard and keeps the implementation simpler.

## Avoid

- Do not return a truncated upstream response. That corrupts JSON, compressed bodies, and any response with application-level framing.
- Do not silently drop the body and return success. The sandbox should see a clear rewrite failure rather than malformed upstream data.
- Do not add per-connector limits or environment configuration in the first fix. We do not have enough product signal to justify that complexity.
- Do not add broad defensive error handling around the existing caller. The caller already has a meaningful failure path for forwarding exceptions.
- Do not attempt the larger streaming-rewrite architecture in this issue. That would require revisiting mitmproxy connection timing, secret resolution, response forwarding, and usage tracking. The bounded full-response fix is the right near-term boundary.

## Test coverage

Add low-level `auth_base_forwarder` coverage for:

- A body exactly at the limit is accepted.
- A body one byte over the limit raises the specific too-large exception.
- The bounded read requests `limit + 1` bytes instead of using unbounded `read()`.
- Response and connection cleanup still happens on the over-limit failure path.
- `forward_request()` enforces the auth.base forwarding concurrency cap around the `to_thread` call.

Existing `auth.py` coverage already verifies that forwarding exceptions become a 502 `url_rewrite_forward_failed` response without marking the rewrite as successful.

### PR DESCRIPTION
**Pull request by seven332**
## Summary
- Bound auth.base forwarded upstream response reads to 32 MiB plus one sentinel byte
- Limit concurrent auth.base forwarding work to four in-flight sync forwards
- Add focused tests for response size boundaries, cleanup, and concurrency capping

Closes #15065

## Validation
- python3 -m pytest tests/test_auth_base_forwarder.py
- python3 -m pytest
- PATH="$HOME/.local/bin:$PATH" ruff check src/auth_base_forwarder.py tests/test_auth_base_forwarder.py
- PATH="$HOME/.local/bin:$PATH" ruff format --check src/auth_base_forwarder.py tests/test_auth_base_forwarder.py
- python3 -m py_compile src/auth_base_forwarder.py tests/test_auth_base_forwarder.py
- git diff --check

### CODE REVIEW

File: crates/runner/mitm-addon/src/auth_base_forwarder.py [hunk 1]
```diff
@@ -183,19 +199,31 @@ def _forward_request_sync(
...
 
 
+def _get_forward_request_semaphore() -> asyncio.Semaphore:
+    global _forward_request_semaphore
+    global _forward_request_semaphore_loop
+
+    loop = asyncio.get_running_loop()
+    if _forward_request_semaphore is None or _forward_request_semaphore_loop is not loop:
+        _forward_request_semaphore = asyncio.Semaphore(MAX_CONCURRENT_AUTH_BASE_FORWARDS)
+        _forward_request_semaphore_loop = loop
```
[the remark this answers is not in the document]
Comment (seven332) [reply crates/runner/mitm-addon/src/auth_base_forwarder.py]: Fixed in b5f4b9395 by grouping the event loop and semaphore into a single forward semaphore state tuple.

**Comment by seven332** [comment]
## Code Review: PR #15074

### Summary

Reviewed the auth.base forwarder response-size cap and concurrency limiting changes. The implementation addresses the unbounded response read without adding connector-specific configuration or streaming complexity, and the tests cover the security-relevant edge cases.

### Key Findings

#### Critical Issues (P0)

None.

#### High Priority (P1)

None.

### Testing Review

#### Coverage

- fix: bound auth.base forwarded response reads to 32 MiB plus one sentinel byte -> covered in `crates/runner/mitm-addon/tests/test_auth_base_forwarder.py`
- fix: cap concurrent auth.base forwarding work at four in-flight requests -> covered in `crates/runner/mitm-addon/tests/test_auth_base_forwarder.py`
- regression: forwarding errors release the semaphore slot -> covered in `crates/runner/mitm-addon/tests/test_auth_base_forwarder.py`

#### Convention Compliance

No blocking convention violations found. The tests use the existing mitm-addon pytest style with targeted Python patching at the low-level HTTP boundary. The direct internal assertions are acceptable for this security-critical forwarder because the behavior under review is resource limiting and cleanup.

#### Testing Verdict: Adequate

### Bad Smell Analysis

- Critical issues: 0
- High-priority issues: 0
- Lint/type suppressions: 0
- Dynamic imports: 0
- TypeScript `any`: not applicable
- Hardcoded environment URLs: 0
- Defensive error handling introduced: 0
- Testing convention blockers: 0

### Recommendations

No changes requested. The remaining slow-upstream risk is bounded by the existing connection timeout and the new four-forward concurrency cap; any stricter total response deadline can be tracked separately if product requirements call for it.

### Verdict

LGTM

---
*Full review details: `codereviews/20260527/pr-15074-*`*
*Testing standards: `docs/testing.md`*
